### ISSUE TITLE
not KL/reverse-KL loss for on policy distillation

### ISSUE THREAD

**Issue by whq-pku** [THUDM/slime#1068]
when I read the implementation of on policy distillation based on slime, the loss function is not a KL loss. Concretely, the advantage is defined as:
https://github.com/THUDM/slime/blob/fcf4e6a2e54f2541f1fa62a3ab982115e12ae72f/slime/backends/megatron_utils/loss.py#L296
with no gradient. and then the loss is defined as:
https://github.com/THUDM/slime/blob/fcf4e6a2e54f2541f1fa62a3ab982115e12ae72f/slime/utils/ppo_utils.py#L133
where `ratio` is -P_s/P_ref. The loss increase P_s when P_t>P_s and decrease P_s when P_t<P_s. It is reasonable. But the supervised information it provides is not as dense as that provided by KL divergence. If we calculate the KL divergence, we need to pass the teacher_logits by `rollout_data`. Why didn't do that? Are there any concerns?

**Comment by ahxt** [comment]
Thanks! We largely follow the implementation described in https://thinkingmachines.ai/blog/on-policy-distillation/, and our experiments on current implmentation also show that this approach works well.

**Comment by Unfinito** [comment]
Your concern is reasonable. These values are not the true KL divergence, but rather an approximation. You can find more details on how this approximation works via sampling (Monte Carlo) in [John Schulman’s blog](http://joschu.net/blog/kl-approx.html).

However, it's crucial to remember that this isn't the true KL because the approximation relies solely on the Top-1 logits (which are primarily used for SFT). Therefore, this method is suitable only for loose control and should not be used for critical tasks.

If you need an accurate KL calculation, you could extract the full logits which you mentioned or the Top-X logits(to save memory) and then apply the standard KL formula to compute it.

**Comment by songmzhang** [comment]
@whq-pku @Unfinito Welcome to try our [KDFlow](https://github.com/songmzhang/KDFlow), an efficient LLM distillation framework that implements lossless KL divergence and many other divergences for LLM on/off-policy distillation.
